### DISCUSSION TITLE
Bromite for windows!!

### DISCUSSION THREAD

**Discussion by uazo** [bromite/bromite#2383] [Builds]
@csagan5 look what I managed to do!

![image](https://user-images.githubusercontent.com/29201891/194538123-6a70a2c9-b631-4d89-a7c3-73154b7a997f.png)
![image](https://user-images.githubusercontent.com/29201891/194888844-38b6250d-55f8-410c-8d41-13848ecb1e4b.png)

I had to set up cross-build from linux, and amazingly it works!
some crashes here and there, some modified patches (mine mostly) but it works!

**Comment by peternrdstrm** [comment]
This is amazing!

**Comment by uazo** [comment]
I ask you a courtesy, please don't be offended.

I would like to come out with the windows version and publish it somewhere sooner or later.
I would like to know what you think about the possibility of still calling it bromite, whether you agree or disagree, and what you think the necessary steps might be.
you have to excuse me but in your evaluation you should consider that if we call it bromite (and i really hope it can be) i would like to have more power over the repo, but it can be separate from the android one.
and I am happy to evaluate any rules you would like me to follow.

**Reply by csagan5** [reply]
> I ask you a courtesy, please don't be offended.

I don't follow, offended by ...?

> I would like to come out with the windows version and publish it somewhere sooner or later.
> I would like to know what you think about the possibility of still calling it bromite, whether you agree or disagree, and what you think the necessary steps might be.

I disagree, Windows/Linux versions are interesting but require a different type of work/tuning than the Android version. I rely on https://github.com/ungoogled-software/ungoogled-chromium for that and I simply cannot see the need for a Bromite for Windows.
What would be the benefits of Bromite for Windows versus ungoogled-chromium for Windows?

> you should consider that if we call it bromite (and i really hope it can be) i would like to have more power over the repo, but it can be separate from the android one.
> and I am happy to evaluate any rules you would like me to follow.

Sorry, not possible. I have no interest in making and maintaining official Windows builds; you could share your build process though, so that others can also try (following open source ideals).

**Reply by uazo** [reply]
> I don't follow, offended by ...?

...for the request, but "offence" I guess is not the right term, that's what 'deepl' proposes to me :) for "non te la prendere"

> What would be the benefits of Bromite for Windows versus ungoogled-chromium for Windows?

the same for which bromite exists together with the android version of ungoogle-chromium.
different objectives I imagine: for example, bromite is configured to be immediately usable while in ungoogle-chromium the various privacy control switches have to be modified by hand.
the target audience also, which I think may be less technical and more 'average': I too would like to have only one browser on my PC and not have to switch to another if a site does not work.
the purpose, in principle eliminating google completely should be a user's choice, not the browser's. 
the method, I personally do not like the automatic substitution (not even in bromite to tell the truth), although it may work I prefer to check what I change and where.
security, some patches seem to me to break mechanisms like the configuration hash protection, and on that as soon as I have the assurance that I'm not saying stupid things I'll expose it to them.
however take these considerations with the benefit of the doubt, only lately I am reading the patches in that repo.

I respect their project and think that now that I am able to compile for windows I can provide some help, but I think that as a contributor to bromite the most appropriate choice is a bromite for windows rather than proposing substantial changes to ungoogle-chromium.
the best would be if there was only one browser that included all the changes made by different chromium forks teams, I had hoped to propose this to some charismatic figure but he has been missing lately.

> Sorry, not possible. I have no interest in making and maintaining official Windows builds; you could share your build process though, so that others can also try (following open source ideals).

I am truly sorry. It is not my absolute intention to make yet another fork and contribute to dissipating the few energies available in the non-profit world of open source forks.
if it helps in changing your choice, it was not my intention to ask you for more activity, I would have taken care of the windows version completely, e.g. by differentiating the changes for windows, so that you don't have to weigh yourself more.
would be complex, but not impossible.

**Reply by uazo** [reply]
>  It is not my absolute intention to make yet another fork 

ah, I forgot, of course of the windows version only, it is absolutely obvious to me that there is only one bromite for android

**Reply by kevin0t** [reply]
> the same for which bromite exists together with the android version of ungoogle-chromium.\ndifferent objectives I imagine: for example, bromite is configured to be immediately usable while in ungoogle-chromium the various privacy control switches have to be modified by hand

One con of using ungoogled-chromium on android is that a user can't use addons like u-block origin to prevent fingerprinting attempts , whereas on windows he can use such things.

But i still see value in this project as u said that bromite is by default configured to provide the user atmost privacy.
I have had used ungoogled chromium in past with u-block for a long time , but once i had to reset my pc and after that i didn't bother to go through all that manual work again to make it work according to me.
A chromium based forks providing bromite level easy setup would be definitely useful for me.

I honestly don't care about updates much , but hate when i have to setup the browser again after a new install. and then ending up using brave which i don't feel is ideal.

**Reply by csagan5** [reply]
> A chromium based forks providing bromite level easy setup would be definitely useful for me.

I understand. However it has never been a goal of Bromite to be easy and working "mostly well" for everyone out of the box, simply because I believe that the average user does not exist and there will always be someone left out, thus the possibility of customisation of an user agent (the browser) is a necessary feature. Bromite disregards the customisation freedom only when it would clearly induce user privacy violations (I am referring to a [past WebRTC debacle](https://github.com/ungoogled-software/ungoogled-chromium/pull/1845), and incidentally [this is a recent vulnerability](https://niespodd.github.io/webrtc-local-ip-leak/) that Bromite is immune from thanks to the sane defaults).

I think most of the "works well out of the box" experience is due to the platform being Android: due to its limitations all apps must pick "decent average settings" because the user interface does not allow very dense and acrobatic configuration operations.
When moving to desktop this limitation is no more there so it would correspond to a sensible effort in terms of offering configuration options that are not available (nor needed) on Android. @uazo is also familiar with this, as we have often compared what is possible for desktop vs mobile

> I am truly sorry. It is not my absolute intention to make yet another fork and contribute to dissipating the few energies available in the non-profit world of open source forks.
> if it helps in changing your choice, it was not my intention to ask you for more activity, I would have taken care of the windows version completely, e.g. by differentiating the changes for windows, so that you don't have to weigh yourself more.
> would be complex, but not impossible.

The only way to make it official would be that I can build it on the same boxes where official Bromite is built and sign it accordingly.

> the same for which bromite exists together with the android version of ungoogle-chromium.
> different objectives I imagine: for example, bromite is configured to be immediately usable while in ungoogle-chromium the various privacy control switches have to be modified by hand.
> the target audience also, which I think may be less technical and more 'average': I too would like to have only one browser on my PC and not have to switch to another if a site does not work.

I think you are missing a big fact here: ungoogled-chromium has been making builds for Windows/Linux for many years, and that means that there is there a good know-how about it and also a lot of patches that are not present in Bromite. I would not dismiss this and consider that Bromite with current patches is a good choice for desktop; I think it would be just a little bit better than Chromium, until some research about what's missing is done.

> security, some patches seem to me to break mechanisms like the configuration hash protection, and on that as soon as I have the assurance that I'm not saying stupid things I'll expose it to them.
> however take these considerations with the benefit of the doubt, only lately I am reading the patches in that repo.

The fingerprinting mitigations? We have options also in Bromite for that, which were sourced from ungoogled-chromium's patches.

**Reply by uazo** [reply]
> The only way to make it official would be that I can build it on the same boxes where official Bromite is built and sign it accordingly.

we can try, but I don't know your build environment.
i followed the instructions for setup the [cross build in linux](https://chromium.googlesource.com/chromium/src.git/+/HEAD/docs/win_cross.md#if-you_re-not-at-google), and added a [casefold disk](https://unix.stackexchange.com/questions/558977/how-to-enable-new-in-kernel-5-2-case-insensitivity-for-ext4-on-a-given-directory) for the windows sdk, i can follow you on the setup, it will take us a while.
first check whether your kernel supports `casefold (+F) flag` and whether you can add other disks to the virtual machine, otherwise it is a big problem.

> until some research about what's missing is done.

I agree, I would start with this

> The fingerprinting mitigations? 

no, some risky code removal used to create a hash vector to limit changes to configurations from external applications.
I think it was also related to the portable mode, but I don't remember well now.

**Comment by SyntheticBird45** [comment]
Hi i'm not particularly legitimate or knowledgeable to ask for a desktop version, but i want to tell you about my situation.

I've always used Bromite on my phone because i was more comfortable with chromium. When i switched to GrapheneOS, vanadium was pre-installed. I quickly understood that security was a serious thing and the difference in philosophy. 

Fortunately, i've learned how insecure is Firefox on both mobile (disastreous) and desktop (especially on Linux). A lot of users now have already saw this https://madaidans-insecurities.github.io/firefox-chromium.html.
But not only, you just have to look at some security researchers on Youtube that share something on their screen to see Chromium somewhere.

But here's the issue. IF you want to take security seriously (which anyone should do), you have to choose between the current chromium forks out there. The first solution is obviously Chromium/Chrome itself, by using some args and flags you can actually significantly reduce the connections being made, and enhance security/privacy (see https://github.com/dreammjow/ChromiumHardening, even if it is archived). some "security people" have also propose to use Edge. Which is not a dumb at all, JIT-disabling, Better sandboxing on Windows and other great security additions, a twitter user even found an undocumented argument to disable telemetry. There is also Brave... but you know it's Brave and it's shady crypto trend. And finally you have ungoogled-chromium. A lot of people have chosen to use ungoogled-chromium for protecting their privacy. but this : https://qua3k.github.io/ungoogled/

So... imagine my situation, where i can't use :
- Edge, i don't think i need to argue here
- Chromium, because using flags and args are not sufficient to stop home telemetry
- Brave, because it is not completely anonymous to use and secure.
- ungoogled-chromium, because of their philosophy (completely insecure).

My actual solution has been to compile chromium with the Hexavalent-Project patches, some ungoogled-chromium patches, release mode, stripping symbols, compiling with clang-cfi  and pairing this with flags. It was actually simple because of Arch Build System, but compiling chromium with these patches at every release is tiring (takes about 2.3 hours) and i've no guarantee that these patch are going to work forever.

That's why i'm really excited of what @uazo have achieved! Please, for the sake of people who believe in both security & privacy, bring us Bromite on desktop (it will only deserve love).

Also, @uazo  if you could send the modified patches in a repository or somewhere i'll be very glad, because i'm struggling apdating patch command lines :)

**Reply by uazo** [reply]
thanks, the links you sent seem interesting.

> IF you want to take security seriously

first of all: for me this is a hobby, so the most trivial answer is 'no, I don't take security seriously', it's not my job, I don't get paid for it.
and I would also say the same for privacy. then, that I am interested, certainly, that there is so much to learn, even more.
carl is better at making choices than I am, and his advice is valuable: I am more of a 'let's try and see'.

so if your goal is absolutely safety, don't use anything I produce. my goal is simply to provide my friends (and me) with a fully functional browser with less privacy (and perhaps security) issues. that's all. and meanwhile I am enjoying myself and I learn new things (and hopefully also the English language :).

> you have to choose between the current chromium forks out there

you forgot about bromite :)

> My actual solution has been to compile chromium with the Hexavalent-Project patches, some ungoogled-chromium patches, release mode, stripping symbols, compiling with clang-cfi and pairing this with flags

you are welcome! i still haven't managed to activate cfi in the windows build, it crashes at the start, how did you do it?

> if you could send the modified patches in a repository

all public https://github.com/uazo/bromite/tree/next-v108 intentionally not organised, bromite is this repo, not mine.
https://github.com/uazo/bromite-buildtools/tree/master/.github/workflows my build process.

**Reply by csagan5** [reply]
> I've always used Bromite on my phone because i was more comfortable with chromium. When i switched to GrapheneOS, vanadium was pre-installed. I quickly understood that security was a serious thing and the difference in philosophy.

GrapheneOS/Vanadium is still probably the best choice for security, however their attitude with this project has been "weird" (you can make an opinion yourself, I will not comment further) and I do not consider them open-source friendly. This might eventually have an impact on the health/quality of the project, but for the maximum benefit of the users I wish it to never happen.



> Fortunately, i've learned how insecure is Firefox on both mobile (disastreous) and desktop (especially on Linux). A lot of users now have already saw this https://madaidans-insecurities.github.io/firefox-chromium.html.

I replied to this when the post was still one of the initial versions; I see this page has changed a lot since then, it seems better now. I can't imagine the considerable work needed to keep an up-to-date list of vulnerabilities for each browser, but you shouldn't rely on a blog post for that, it is a very dynamic situation changing day by day.



> But here's the issue. IF you want to take security seriously (which anyone should do), you have to choose between the current chromium forks out there.

A vendor-provided browser (with a company behind it) is the best choice for security. Full stop. Individuals or small open source projects do not have the resources to provide anyone with a browser as much secure as that.



> There is also Brave... but you know it's Brave and it's shady crypto trend.

Your evaluation should be based on security, not on rumours. I do not use Brave but I invite everyone to discuss security by talking about...security.



> A lot of people have chosen to use ungoogled-chromium for protecting their privacy. but this : https://qua3k.github.io/ungoogled/

If your method is "let's find a browser that has no critiques against it", you are in for a sour surprise. No browser is perfect or perfectly secure, and you cannot objectively compare them to each other in terms of security. You could count CVEs, or you could estimate vulnerabilities impact, but it all depends also from the way an user uses the browser.



> That's why i'm really excited of what @uazo have achieved! Please, for the sake of people who believe in both security & privacy, bring us Bromite on desktop (it will only deserve love).

That is until someone makes a comprehensive list of the problems of Bromite (on desktop); nothing against that, although I would prefer that they are also submitted as issues/discussions here to this project rather than on blog posts, for the sake of improving the project.
